warehouse/query-core/src/main/java/datawave/query/iterator/GroupingIterator.java

```diff
@@ -0,0 +1,240 @@
+import org.apache.accumulo.core.security.ColumnVisibility;
+import org.slf4j.Logger;
+import org.springframework.util.Assert;
+
+import java.math.BigDecimal;
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * Because the t-server may tear down and start a new iterator at any time after a next() call, there can be no saved state in this class. For that reason, each
+ * next call on the t-server will flatten the aggregated data into a single Entry&gt;Key,Document&lt; to return to the web server.
+ */
+public class GroupingIterator implements Iterator<Map.Entry<Key,Document>> {
+    
+    private static final Logger log = getLogger(GroupingIterator.class);
+    
+    /**
+     * the fields (user provided) to group by
+     */
+    private final Set<String> groupFieldsSet;
+    
+    /**
+     * A map of TypeAttribute collection keys to integer counts This map uses a special key type that ignores the metadata (with visibilities) in its hashCode
+     * and equals methods
+     */
+    private GroupCountingHashMap countingMap;
+    
+    /**
+     * holds the aggregated column visibilities for each grouped event
+     */
+    private Multimap<Collection<GroupingTypeAttribute<?>>,ColumnVisibility> fieldVisibilities = HashMultimap.create();
+    
+    private final GroupingUtil groupingUtil = new GroupingUtil();
+    
+    /**
+     * list of keys that have been read, in order to keep track of where we left off when a new iterator is created
+     */
+    private final List<Key> keys = new ArrayList<>();
+    
+    private final MarkingFunctions markingFunctions;
+    
+    private final int groupFieldsBatchSize;
+    
+    private final YieldCallback<Key> yieldCallback;
+    
+    private final Iterator<Map.Entry<Key,Document>> previousIterators;
+    
+    private final LinkedList<Document> documents = new LinkedList<>();
+    
+    Map.Entry<Key,Document> next;
+    
+    public GroupingIterator(Iterator<Map.Entry<Key,Document>> previousIterators, MarkingFunctions markingFunctions, Collection<String> groupFieldsSet,
+                    int groupFieldsBatchSize, YieldCallback<Key> yieldCallback) {
+        this.previousIterators = previousIterators;
+        this.markingFunctions = markingFunctions;
+        this.groupFieldsSet = groupFieldsSet.stream().map(JexlASTHelper::deconstructIdentifier).collect(Collectors.toSet());
+        this.groupFieldsBatchSize = groupFieldsBatchSize;
+        this.yieldCallback = yieldCallback;
+        
+        this.countingMap = new GroupCountingHashMap(this.markingFunctions);
+    }
+    
+    @Override
+    public boolean hasNext() {
+        for (int i = 0; i < groupFieldsBatchSize; i++) {
+            if (previousIterators.hasNext()) {
+                Map.Entry<Key,Document> entry = previousIterators.next();
+                if (entry != null) {
+                    log.trace("{} get list key counts for: {}", "t-server", entry);
```

Comment:
Didn't mean to. This was a copy/paste and then modified from the code before it was separated into GroupingIterator and GroupingTransform (it used to be a ternary statement to substitute either t-server or webclient) -- changed to use "t-server get list...." instead of substitution.